### PR TITLE
Outbox publisher (GSI-773)

### PR DESCRIPTION
**Pull request by TheByronHimes**
This PR replaces the normal event publisher with the newer outbox dao publisher. When a file ID is specified for deletion, the File Deletion core layer inserts the file ID into the database and immediately publishes the data as an outbox-style event (meaning the extra `__metadata__` stored in the mongodb document is added to the kafka event). 

This ensures that requests are not lost if something happens to kafka -- the file deletion request events can be republished from the database.

Tests were slightly refactored and expanded to include testing the deletion functionality of the DAO publisher, even though that is not used at this time.

### CODE REVIEW

File: services/pcs/tests_pcs/fixtures/joint.py [hunk 1]
```diff
@@ -22,53 +22,54 @@
...
-                yield JointFixture(
-                    config=config,
-                    file_deletion=file_deletion,
-                    rest_client=rest_client,
-                    kafka=kafka,
-                    token=token,
-                )
+    config = get_config(sources=[mongo_kafka.config, kafka.config, token_hash_config])
+    async for dao in get_file_deletion_dao(config=config):
+        async with prepare_core(config=config) as file_deletion:
```
Comment (mephenor) [inline services/pcs/tests_pcs/fixtures/joint.py:62]:
```suggestion
    async with (get_file_deletion_dao(config=config) as dao, prepare_core(config=config) as file_deletion):
```
Needs some formatting, but should work the same as now and the intent with the dao is clear, the current `for` is quite confusing

File: services/pcs/tests_pcs/test_typical_journey.py [hunk 2]
```diff
@@ -17,46 +17,94 @@
...
 
+pytestmark = pytest.mark.asyncio()
 
-@pytest.mark.asyncio
-async def test_journey(joint_fixture: JointFixture):  # noqa: F811
-    """Simulates a typical, successful API journey."""
-    file_id = "test_id"
 
-    files_deletion_event = event_schemas.FileDeletionRequested(file_id=file_id)
+test_file_id = "test_id"
```
Comment (mephenor) [inline services/pcs/tests_pcs/test_typical_journey.py:35]:
```suggestion
TEST_FILE_ID = "test_id"
```

File: services/pcs/src/pcs/inject.py [hunk 3]
```diff
@@ -21,22 +21,42 @@
...
+
+
+async def get_mongo_kafka_dao_factory(
+    config: Config,
+) -> AsyncGenerator[MongoKafkaDaoPublisherFactory, None]:
+    """Get a MongoDB DAO publisher factory."""
+    async with MongoKafkaDaoPublisherFactory.construct(config=config) as factory:
+        yield factory
+
+
```
Comment (mephenor) [inline services/pcs/src/pcs/inject.py:42]:
```suggestion
@asynccontextmanager
```

File: services/pcs/src/pcs/inject.py [hunk 4]
```diff
@@ -21,22 +21,42 @@
...
+        yield file_deletion_dao
 
 
 @asynccontextmanager
 async def prepare_core(*, config: Config) -> AsyncGenerator[FileDeletionPort, None]:
     """Construct and initialize the core component and its outbound dependencies."""
-    async with KafkaEventPublisher.construct(config=config) as event_pub_provider:
-        event_publisher = EventPubTranslator(config=config, provider=event_pub_provider)
-        file_deletion = FileDeletion(event_publisher=event_publisher)
+    async for file_deletion_dao in get_file_deletion_dao(config=config):
```
Comment (mephenor) [inline services/pcs/src/pcs/inject.py:58]:
```suggestion
    async with get_file_deletion_dao(config=config) as file_deletion_dao:
```

File: services/pcs/src/pcs/adapters/outbound/daopub.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        """Configure with provider for the DaoFactoryProtocol"""
+        self._dao_publisher_factory = dao_publisher_factory
+        self._file_deletions_collection = config.file_deletions_collection
+        self._file_deletion_topic = config.files_to_delete_topic
+
+    @staticmethod
+    def _file_id_to_event(file_deletion_event: FileDeletionRequested) -> JsonObject:
+        """Translate a user to an event."""
+        validated_model = FileDeletionRequested(file_id=file_deletion_event.file_id)
+        return validated_model.model_dump()
```
Comment (mephenor) [inline services/pcs/src/pcs/adapters/outbound/daopub.py:66]: Couldn't you just directly `model_dump()` the event model instead?
Seems like unnecessary indirection

**Review by mephenor — CHANGES_REQUESTED** [review]
A few smaller issues, with one that needs changes in a few places:
The `get_file_deletion` dao can be converted to an async context manager for its usage to be in line with the other functionality in the joint fixture

File: services/pcs/tests_pcs/test_typical_journey.py (hunk 2 above)
Comment (TheByronHimes) [reply services/pcs/tests_pcs/test_typical_journey.py]: Forgot about the upper case, thanks

File: services/pcs/src/pcs/inject.py (hunk 4 above)
Comment (TheByronHimes) [reply services/pcs/src/pcs/inject.py]: You're right, that is better.

File: services/pcs/src/pcs/adapters/outbound/daopub.py (hunk 5 above)
Comment (TheByronHimes) [reply services/pcs/src/pcs/adapters/outbound/daopub.py]: Yes, I replaced it with a lambda. Not sure why I did that, maybe changed things around and just didn't catch it.

**Review by mephenor — APPROVED** [review]

**Pull request merged**